compiler/src/dotty/tools/dotc/transform/init/Objects.scala

```diff
@@ -0,0 +1,1362 @@
+      // calls and user code in the class body.
+      mixins.reverse.foreach { mixin =>
+        parents.find(_.tpe.classSymbol == mixin) match
+        case Some(parent) =>
+          extendTrace(parent) { initParent(parent, tasks) }
+        case None =>
+          // According to the language spec, if the mixin trait requires
+          // arguments, then the class must provide arguments to it explicitly
+          // in the parent list. That means we will encounter it in the Some
+          // branch.
+          //
+          // When a trait A extends a parameterized trait B, it cannot provide
+          // term arguments to B. That can only be done in a concrete class.
+          val tref = typeRefOf(klass.typeRef.baseType(mixin).typeConstructor)
+          val ctor = tref.classSymbol.primaryConstructor
+          if ctor.exists then
+            // The parameter check of traits comes late in the mixin phase.
+            // To avoid crash we supply hot values for erroneous parent calls.
+            // See tests/neg/i16438.scala.
+            val args: List[ArgInfo] = ctor.info.paramInfoss.flatten.map(_ => new ArgInfo(Bottom, Trace.empty))
+            extendTrace(superParent) {
+              superCall(tref, ctor, args, tasks)
+            }
+      }
+
+      // initialize super classes after outers are set
+      tasks.foreach(task => task())
+    end if
+
+    // class body
+    tpl.body.foreach {
+      case vdef : ValDef if !vdef.symbol.is(Flags.Lazy) && !vdef.rhs.isEmpty =>
+        val res = eval(vdef.rhs, thisV, klass)
+        val sym = vdef.symbol
+        if sym.is(Flags.Mutable) then
+          val addr = Heap.fieldVarAddr(summon[Regions.Data], sym, State.currentObject)
+          thisV.initVar(sym, addr)
+          Heap.write(addr, res)
+        else
+          thisV.initVal(sym, res)
+
+      case _: MemberDef =>
+
+      case tree =>
+        eval(tree, thisV, klass)
+    }
+
+    thisV
+  }
+
+
+  /** Resolve C.this that appear in `klass`
+   *
+   * @param target  The class symbol for `C` for which `C.this` is to be resolved.
+   * @param thisV   The value for `D.this` where `D` is represented by the parameter `klass`.
+   * @param klass   The enclosing class where the type `C.this` is located.
+   * @param elideObjectAccess Whether object access should be omitted.
+   *
+   * Object access elission happens when the object access is used as a prefix
+   * in `new o.C` and `C` does not need an outer.
+   */
+  def resolveThis(target: ClassSymbol, thisV: Value, klass: ClassSymbol, elideObjectAccess: Boolean = false): Contextual[Value] = log("resolveThis target = " + target.show + ", this = " + thisV.show, printer, (_: Value).show) {
+    if target == klass then
+      thisV
+    else if target.is(Flags.Package) then
+      Bottom
+    else if target.isStaticObject then
+      val res = ObjectRef(target.moduleClass.asClass)
+      if target == klass || elideObjectAccess then res
```

Comment:
`target == klass` can't happen because there is a case for it earlier.

Reply:
Good catch 👍